odoo/fields.py

```diff
@@ -2821,11 +2846,18 @@ def convert_to_write(self, value, record):
                 else:
                     result[0][2].append(origin.id)
                     if record != origin:
-                        values = record._convert_to_write({
-                            name: record[name]
-                            for name in record._cache
-                            if name not in inv_names and record[name] != origin[name]
-                        })
+                        values = dict()
+                        for fname in record._cache:
+                            if fname in inv_names:
+                                continue
+                            if record[fname] == origin[fname]:
+                                continue
+                            field = record._fields[fname]
+                            val = field.convert_to_write(record[fname], record)
+                            if not record.id and field.type in ("one2many", "many2many")\
+                                and len(val) == 1 and val[0][0] == 6 and val[0][2] == origin[fname].ids:
```

Comment:
It would be easier if `convert_to_write()` returns `[]` when nothing has changed in the x2many...

Reply:
No, because it means we would put an empty `[ ]` in `values[fname]`, an a `(1, 0, {fname:[ ]}` would still be added to the returned result, bringing back the same problem of writing for no effective change.

I added a test to cover the case.